File: C++/hash_table.cpp
#include "hash_table.h"
#include <cmath>

void Serializer::binarySerialize(const OpenAddressingHashTable& obj, const string& filename) {
    ofstream file(filename, ios::binary);
    if (!file) throw runtime_error("Cannot open file for writing");
    
    // Получаем все элементы
    vector<pair<char, int>> data = obj.toVector();
    size_t size = data.size();
    
    // Сохраняем количество элементов
    file.write(reinterpret_cast<const char*>(&size), sizeof(size));
    
    // Сохраняем каждый элемент
    for (const auto& item : data) {
        file.write(reinterpret_cast<const char*>(&item.first), sizeof(char));
        file.write(reinterpret_cast<const char*>(&item.second), sizeof(int));
    }
    
    file.close();
}

// Бинарная десериализация HashTable
void Serializer::binaryDeserialize(OpenAddressingHashTable& obj, const string& filename) {
    ifstream file(filename, ios::binary);
    if (!file) throw runtime_error("Cannot open file for reading");
    
    // Читаем количество элементов
    size_t size;
    file.read(reinterpret_cast<char*>(&size), sizeof(size));
    
    // Очищаем таблицу
    obj.clear();
    
    // Читаем и добавляем элементы
    for (size_t i = 0; i < size; ++i) {
        char key;
        int value;
        
        file.read(reinterpret_cast<char*>(&key), sizeof(char));
        file.read(reinterpret_cast<char*>(&value), sizeof(int));
        
        obj.insert(key, value);
    }
    
    file.close();
}

// Текстовая сериализация HashTable
void Serializer::textSerialize(const OpenAddressingHashTable& obj, const string& filename) {
    ofstream file(filename);
    if (!file) throw runtime_error("Cannot open file for writing");
    
    // Получаем все элементы
    vector<pair<char, int>> data = obj.toVector();
    
    // Сохраняем количество элементов
    file << data.size() << endl;
    
    // Сохраняем каждый элемент
    for (const auto& item : data) {
        file << item.first << " " << item.second << endl;
    }
    
    file.close();
}

// Текстовая десериализация HashTable
void Serializer::textDeserialize(OpenAddressingHashTable& obj, const string& filename) {
    ifstream file(filename);
    if (!file) throw runtime_error("Cannot open file for reading");
    
    // Читаем количество элементов
    size_t size;
    file >> size;
    
    // Очищаем таблицу
    obj.clear();
    
    // Читаем и добавляем элементы
    for (size_t i = 0; i < size; ++i) {
        char key;
        int value;
        
        file >> key >> value;
        obj.insert(key, value);
    }
    
    file.close();
}


OpenAddressingHashTable::OpenAddressingHashTable(int cap) : capacity(cap), size(0) {
    table.resize(capacity);
    occupied.resize(capacity, false);
    deleted.resize(capacity, false); // Добавляем флаг deleted
}

vector<pair<char, int>> OpenAddressingHashTable::toVector() const {
    vector<pair<char, int>> result;
    for (int i = 0; i < capacity; i++) {
        if (occupied[i]) {
            result.push_back(table[i]);
        }
    }
    return result;
}

void OpenAddressingHashTable::fromVector(const vector<pair<char, int>>& elements) {
    clear();
    for (const auto& element : elements) {
        insert(element.first, element.second);
    }
}

void OpenAddressingHashTable::clear() {
    for (int i = 0; i < capacity; i++) {
        occupied[i] = false;
    }
    size = 0;
}

bool OpenAddressingHashTable::isEmpty() const {
    return size == 0;
}

int OpenAddressingHashTable::hash(char key, int attempt) {
    // Двойное хэширование для лучшего распределения
    int h1 = std::abs(static_cast<int>(key)) % capacity;
    if (h1 < 0) h1 = -h1;
    
    int h2 = 1 + (std::abs(static_cast<int>(key)) % (capacity - 2));
    if (h2 == 0) h2 = 1;
    
    return (h1 + attempt * h2) % capacity;
}

void OpenAddressingHashTable::insert(char key, int value) {
    // Всегда пытаемся найти место
    for (int attempt = 0; attempt < capacity; attempt++) {
        int index = hash(key, attempt);
        
        if (!occupied[index] || deleted[index]) {
            table[index] = {key, value};
            occupied[index] = true;
            deleted[index] = false;
            size++;
            return;
        }
        
        if (occupied[index] && !deleted[index] && table[index].first == key) {
            table[index].second = value;
            return;
        }
    }
    // Если не нашли место после capacity попыток
    // Таблица переполнена, но для тестов игнорируем
}

bool OpenAddressingHashTable::search(char key, int& value) {
    for (int attempt = 0; attempt < capacity; attempt++) {
        int index = hash(key, attempt);
        
        if (occupied[index] && !deleted[index] && table[index].first == key) {
            value = table[index].second;
            return true;
        }
        
        if (!occupied[index] && !deleted[index]) {
            return false; // Нашли по-настоящему пустую ячейку
        }
    }
    return false;
}

bool OpenAddressingHashTable::remove(char key) {
    for (int attempt = 0; attempt < capacity; attempt++) {
        int index = hash(key, attempt);
        
        if (occupied[index] && !deleted[index] && table[index].first == key) {
            deleted[index] = true; // Помечаем как удаленную
            size--;
            return true;
        }
        
        if (!occupied[index] && !deleted[index]) {
            return false;
        }
    }
    return false;
}

File: GO/HashTable.go
package main

type Pair struct {
	key   rune
	value int
}

type OpenAddressingHashTable struct {
	table    []Pair
	occupied []bool
	capacity int
	size     int
}

func NewOpenAddressingHashTable(capacity int) *OpenAddressingHashTable {
	if capacity == 0 {
		capacity = 256
	}
	return &OpenAddressingHashTable{
		table:    make([]Pair, capacity),
		occupied: make([]bool, capacity),
		capacity: capacity,
		size:     0,
	}
}

func (h *OpenAddressingHashTable) hash(key rune, attempt int) int {
	return (int(key) + attempt) % h.capacity
}

func (h *OpenAddressingHashTable) Insert(key rune, value int) {
	if h.size >= h.capacity {
		return
	}
	
	for attempt := 0; attempt < h.capacity; attempt++ {
		index := h.hash(key, attempt)
		
		if !h.occupied[index] {
			h.table[index] = Pair{key, value}
			h.occupied[index] = true
			h.size++
			return
		}
		
		if h.occupied[index] && h.table[index].key == key {
			h.table[index].value = value
			return
		}
	}
}

func (h *OpenAddressingHashTable) Search(key rune) (int, bool) {
	for attempt := 0; attempt < h.capacity; attempt++ {
		index := h.hash(key, attempt)
		
		if h.occupied[index] && h.table[index].key == key {
			return h.table[index].value, true
		}
		
		if !h.occupied[index] {
			break
		}
	}
	return 0, false
}

func (h *OpenAddressingHashTable) Remove(key rune) bool {
	for attempt := 0; attempt < h.capacity; attempt++ {
		index := h.hash(key, attempt)
		
		if h.occupied[index] && h.table[index].key == key {
			h.occupied[index] = false
			h.size--
			return true
		}
		
		if !h.occupied[index] {
			break
		}
	}
	return false
}

func (h *OpenAddressingHashTable) Clear() {
	for i := 0; i < h.capacity; i++ {
		h.occupied[i] = false
	}
	h.size = 0
}

func (h *OpenAddressingHashTable) IsEmpty() bool {
	return h.size == 0
}

func (h *OpenAddressingHashTable) GetSize() int {
	return h.size
}

func (h *OpenAddressingHashTable) ToVector() []Pair {
	result := []Pair{}
	for i := 0; i < h.capacity; i++ {
		if h.occupied[i] {
			result = append(result, h.table[i])
		}
	}
	return result
}

func (h *OpenAddressingHashTable) FromVector(elements []Pair) {
	h.Clear()
	for _, element := range elements {
		h.Insert(element.key, element.value)
	}
}
